[feature] Implemented view permissions by extending `DjangoModelPermissions` class #249

ManishShah120 opened the pull request:
Closes #249

ManishShah120 commented on openwisp_users/api/permissions.py:
```diff
@@ -66,3 +68,32 @@ class IsOrganizationOwner(BaseOrganizationPermission):
+        perms = self.get_required_permissions(request.method, queryset.model)
+        change_perm = self.get_required_permissions('PUT', queryset.model)
+
+        if request.method == 'GET':
+            if request.user.has_perms(perms) or request.user.has_perms(change_perm):
+                return True
+            else:
+                return False
```
Did not handled the case of super user becasue superuser by default has all the permissions.

atb00ker replied:
I think you can add a check at the top stating:
```
if user.is_superuser:
    return True
```

They are allowed to do anything regardless of any permission!

ManishShah120 replied:
@atb00ker I am trying to implement this in more generic way, cause am thinking of opening a similar PR in DRF repo once this gets merged here. 

By default `superuser` has all permissions i.e., it can do anything it wants. suppose manually someone removes all the permissions for any specific model for `superuser`. Then in that case this would not work, since we will return True if it;s a super user, so that case will fail.

What do you think?

atb00ker replied:
To the best of my knowledge, removing permissions from `superuser` is an anti-pattern, but I understand your point.
1. We should look at existing DRF functions and implement it in a similar way.
2. I'll also wait for @nemesisdesign check this! :smile:

atb00ker commented on tests/testapp/tests/test_permission_classes.py:
```diff
@@ -117,3 +121,29 @@ def test_organization_field_with_errored_parent(self):
             self.client.get(reverse('test_error_field_view'), **auth)
         self.assertIn('Organization not found', str(error.exception))
+
+    def test_custom_django_model_permission_with_view_permission(self):
```
Can we have a test for negative case when user does not have the permission; they are given 403.

ManishShah120 replied:
I have updated the tests. :+1:

atb00ker commented on openwisp_users/api/permissions.py:
```diff
@@ -66,3 +68,32 @@ class IsOrganizationOwner(BaseOrganizationPermission):
         return org and (user.is_superuser or user.is_owner(org))
+
+
+class CustomDjangoModelPermissions(DjangoModelPermissions):
```
Just wondering, can we name it something like `ViewDjangoModelPermissions` so that the name makes more sense in the other modules about that's "Custom" about this class!

ManishShah120 replied:
Done @atb00ker :+1:

atb00ker reviewed (commented):
This looks good to me, small comments:

ManishShah120 commented:
Quoting from the issue #249

> **Tests:**

> - ensure an user of the operator group (not flagged as supersuser) cannot view both endpoints
  - Added
> - ensure an user of the administrator group (not flagged as supersuser) can view both endpoints
  - Added
> - add the view permission to the operator group, ensure the user now can view both endpoints but cannot change anything (put or delete)
  - Added

nemesisdesign commented on openwisp_users/api/permissions.py:
```diff
@@ -66,3 +68,31 @@ class IsOrganizationOwner(BaseOrganizationPermission):
+class ViewDjangoModelPermissions(DjangoModelPermissions):
+    def __init__(self):
+        self.perms_map = copy.deepcopy(self.perms_map)
+        self.perms_map['GET'] = ['%(app_label)s.view_%(model_name)s']
```
can we do this at class level so it's done only once at startup?

ManishShah120 replied:
Done

nemesisdesign reviewed (commented):
Looks on the right track! I'll need to test it.

But definitely please go ahead with the PR to Django REST Framework, hopefully we can get something similar or better merged there.

okraits reviewed (commented):
Looks fine. No objections from me.

okraits reviewed (commented):
Looks fine.

atb00ker commented on tests/testapp/tests/test_permission_classes.py:
```diff
@@ -117,3 +126,128 @@ def test_organization_field_with_errored_parent(self):
         self.assertIn('Organization not found', str(error.exception))
+
+    def test_view_permission_with_operator(self):
+        user = User.objects.create_user(
+            username='operator', password='tester', email='operator@test.com'
+        )
+        operator_group = Group.objects.filter(name='Operator')
+        user.groups.set(operator_group)
```
Just FYI, I think there are functions for common actions. Like [_create_operator](https://github.com/openwisp/openwisp-users/blob/master/openwisp_users/tests/utils.py#L53).
These should be inherited and used! :smile:

ManishShah120 replied:
Hey, @atb00ker Yes, I tried to use it but with this function, the user gets all the permissions without being in the `operator` group, and here we only needed `view.testapp_template` permission.

atb00ker commented on tests/testapp/tests/test_permission_classes.py:
```diff
@@ -117,3 +126,128 @@ def test_organization_field_with_errored_parent(self):
+        operator_group = Group.objects.filter(name='Operator')
+        user.groups.set(operator_group)
+        org1 = self._get_org()
+        OrganizationUser.objects.create(user=user, organization=org1, is_admin=True)
```
Same here. Find function here: https://github.com/openwisp/openwisp-users/blob/master/openwisp_users/tests/utils.py#L128

atb00ker reviewed (commented):
I think this is looking good, small improvement in the tests.

atb00ker reviewed (dismissed):
Functional Test:
- [x] superuser has access
- [x] non-su without view permission & org manager doesn't have access
- [x] non-su with view permission & not org manager has access but sees nothing
- [x] non-su with view permission & org manager can see temp for their obj. (not shared obj)

nemesisdesign commented on openwisp_users/api/permissions.py:
```diff
@@ -66,3 +66,37 @@ class IsOrganizationOwner(BaseOrganizationPermission):
+            if user.has_perms(perms) or user.has_perms(change_perm):
+                return True
+            else:
+                return False
```
what about:
```python
# user must have either view permission or change permission
return user.has_perms(perms) or user.has_perms(change_perm)
```

nemesisdesign commented on tests/testapp/tests/test_permission_classes.py:
```diff
@@ -117,3 +126,128 @@ def test_organization_field_with_errored_parent(self):
+        self.client.force_login(user)
+        token = self._obtain_auth_token()
+        auth = dict(HTTP_AUTHORIZATION=f'Bearer {token}')
+        t1 = self._create_template(organization=org1)
```
the first lines of each tests look really similar, please try to make a private helper method which prepares the data needed for the test, so we can make the actual test code shorter and easier to read.

ManishShah120 replied:
Done, Please have a look. :+1:

nemesisdesign commented on openwisp_users/api/permissions.py:
```diff
@@ -66,3 +66,37 @@ class IsOrganizationOwner(BaseOrganizationPermission):
         return org and (user.is_superuser or user.is_owner(org))
+
+
+class ViewDjangoModelPermissions(DjangoModelPermissions):
```
we could call this simply `DjangoModelPermissions`

nemesisdesign commented on openwisp_users/api/permissions.py:
```diff
@@ -1,5 +1,5 @@
 from django.utils.translation import gettext_lazy as _
-from rest_framework.permissions import BasePermission
+from rest_framework.permissions import BasePermission, DjangoModelPermissions
```
here you'd import it as `BaseDjangoModelPermissions`

nemesisdesign reviewed (changes requested):
Great work @ManishShah120 @atb00ker! :+1: 

Missing:

- [ ] Documentation (please add it near to the other sections related to DRF)
- [x] Code simplification (see below)

atb00ker commented on tests/testapp/tests/test_permission_classes.py:
```diff
@@ -117,3 +126,106 @@ def test_organization_field_with_errored_parent(self):
+        return (auth, t1)
+
+    def test_view_permission_with_operator(self):
+        user = self._get_user()
+        operator_group = Group.objects.filter(name='Operator')
+        user.groups.set(operator_group)
```
Doesn't `_get_operator()` help here?

ManishShah120 replied:
No @atb00ker.

atb00ker reviewed (approved):
LGTM! :+1:

nemesisdesign reviewed (approved):
:+1:
